mediapipe/graphs/instantmotiontracking/calculators/gl_animation_overlay_calculator.cc

```diff
@@ -0,0 +1,760 @@
+  GLint perspective_matrix_uniform_ = -1;
+  GLint model_matrix_uniform_ = -1;
+
+  std::vector<TriangleMesh> triangle_meshes_;
+  std::vector<TriangleMesh> mask_meshes_;
+  Timestamp animation_start_time_;
+  int frame_count_ = 0;
+  float animation_speed_fps_;
+
+  std::vector<ModelMatrix> current_model_matrices_;
+  std::vector<ModelMatrix> current_mask_model_matrices_;
+
+  // Perspective matrix for rendering, to be applied to all model matrices
+  // prior to passing through to the shader as a MVP matrix.  Initialized during
+  // first image packet read.
+  float perspective_matrix_[kNumMatrixEntries];
+
+  void ComputeAspectRatioAndFovFromCameraParameters(
+      const CameraParametersProto &camera_parameters, float *aspect_ratio,
+      float *vertical_fov_degrees);
+  int GetAnimationFrameIndex(Timestamp timestamp);
+  ::mediapipe::Status GlSetup();
+  ::mediapipe::Status GlBind(const TriangleMesh &triangle_mesh,
+                             const GlTexture &texture);
+  ::mediapipe::Status GlRender(const TriangleMesh &triangle_mesh,
+                               const float *model_matrix);
+  void InitializePerspectiveMatrix(float aspect_ratio,
+                                   float vertical_fov_degrees, float z_near,
+                                   float z_far);
+  void LoadModelMatrices(const TimedModelMatrixProtoList &model_matrices,
+                         std::vector<ModelMatrix> *current_model_matrices);
+
+#if !defined(__ANDROID__)
+  // Asset loading routine for all non-Android platforms.
+  bool LoadAnimation(const std::string &filename);
+#else
+  // Asset loading for all Android platforms.
+  bool LoadAnimationAndroid(const std::string &filename,
+                            std::vector<TriangleMesh> *mesh);
+  bool ReadBytesFromAsset(AAsset *asset, void *buffer, int num_bytes_to_read);
+#endif
+};
+REGISTER_CALCULATOR(GlAnimationOverlayCalculator);
+
+// static
+::mediapipe::Status GlAnimationOverlayCalculator::GetContract(
+    CalculatorContract *cc) {
+  MP_RETURN_IF_ERROR(
+      GlCalculatorHelper::SetupInputSidePackets(&(cc->InputSidePackets())));
+  if (cc->Inputs().HasTag("VIDEO")) {
+    // Currently used only for size and timestamp.
+    cc->Inputs().Tag("VIDEO").Set<GpuBuffer>();
+  }
+  TagOrIndex(&(cc->Outputs()), "OUTPUT", 0).Set<GpuBuffer>();
+
+  if (cc->Inputs().HasTag("MODEL_MATRICES")) {
+    cc->Inputs().Tag("MODEL_MATRICES").Set<TimedModelMatrixProtoList>();
+  }
+  if (cc->Inputs().HasTag("MASK_MODEL_MATRICES")) {
+    cc->Inputs().Tag("MASK_MODEL_MATRICES").Set<TimedModelMatrixProtoList>();
+  }
+
+  cc->InputSidePackets().Tag("TEXTURE").Set<AssetTextureFormat>();
+  cc->InputSidePackets().Tag("ANIMATION_ASSET").Set<std::string>();
+  if (cc->InputSidePackets().HasTag("CAMERA_PARAMETERS_PROTO_STRING")) {
+    cc->InputSidePackets()
+        .Tag("CAMERA_PARAMETERS_PROTO_STRING")
+        .Set<std::string>();
+  }
+
+  if (cc->InputSidePackets().HasTag("MASK_TEXTURE")) {
```

Comment:
We should have a comment for these up top-- but looks like that was already the case, so no need to do anything at the moment; we should fix that at the source when we get a chance.